### PR TITLE
Fix GetBlockByHash crashed on block not found

### PR DESCRIPTION
**Pull request by calvinaco**
Closes: #255

## Description

Add block not found handling at RPC backend

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: ethereum/rpc/backend/backend.go [hunk 1]
```diff
@@ -129,10 +134,15 @@ func (e *EVMBackend) GetBlockByNumber(blockNum types.BlockNumber, fullTx bool) (
 func (e *EVMBackend) GetBlockByHash(hash common.Hash, fullTx bool) (map[string]interface{}, error) {
 	resBlock, err := e.clientCtx.Client.BlockByHash(e.ctx, hash.Bytes())
 	if err != nil {
-		e.logger.Warningf("BlockByHash failed for %s", hash.Hex())
+		e.logger.WithError(err).Warningf("BlockByHash failed for %s", hash.Hex())
```
Comment (calvinaco) [inline ethereum/rpc/backend/backend.go:137]: @fedekunze I notice here we are using `Warning` log level, wonder if there's any convention when to use warning and debug?

I am leaving the level as it is for now.

File: ethereum/rpc/backend/backend.go [hunk 2]
```diff
@@ -133,6 +137,10 @@ func (e *EVMBackend) GetBlockByHash(hash common.Hash, fullTx bool) (map[string]i
 		return nil, err
 	}
 
+	if resBlock.Block == nil {
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (calvinaco) [reply ethereum/rpc/backend/backend.go:143]: Added, thanks

File: ethereum/rpc/backend/backend.go [hunk 3]
```diff
@@ -117,6 +117,10 @@ func (e *EVMBackend) GetBlockByNumber(blockNum types.BlockNumber, fullTx bool) (
 		}
 	}
 
+	if resBlock.Block == nil {
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (calvinaco) [reply ethereum/rpc/backend/backend.go:122]: Added, thanks

File: ethereum/rpc/backend/backend.go [hunk 4]
```diff
@@ -112,11 +112,16 @@ func (e *EVMBackend) GetBlockByNumber(blockNum types.BlockNumber, fullTx bool) (
...
 		// e.logger.Debugf("GetBlockByNumber safely bumping down from %d to latest", height)
 		if resBlock, err = e.clientCtx.Client.Block(e.ctx, nil); err != nil {
-			e.logger.Debugln("GetBlockByNumber failed to get latest block")
+			e.logger.WithError(err).Debugln("GetBlockByNumber failed to get latest block")
 			return nil, nil
 		}
 	}
 
+	if resBlock.Block == nil {
+		e.logger.Debugf("GetBlockByNumber block not found: %d", height)
```
Comment (fedekunze) [inline ethereum/rpc/backend/backend.go:121]:
```suggestion
		e.logger.Debugln("GetBlockByNumber block not found", "height", height)
```

File: ethereum/rpc/backend/backend.go (hunk 1 above)
Comment (fedekunze) [reply ethereum/rpc/backend/backend.go:137]: yes, I created issue #262 to fix it

Comment (fedekunze) [reply ethereum/rpc/backend/backend.go:137]:
```suggestion
		e.logger.WithError(err).Warningf("BlockByHash block not found", "hash", hash.Hex())
```

File: ethereum/rpc/backend/backend.go [hunk 5]
```diff
@@ -129,10 +134,15 @@ func (e *EVMBackend) GetBlockByNumber(blockNum types.BlockNumber, fullTx bool) (
 func (e *EVMBackend) GetBlockByHash(hash common.Hash, fullTx bool) (map[string]interface{}, error) {
 	resBlock, err := e.clientCtx.Client.BlockByHash(e.ctx, hash.Bytes())
 	if err != nil {
-		e.logger.Warningf("BlockByHash failed for %s", hash.Hex())
+		e.logger.WithError(err).Warningf("BlockByHash failed for %s", hash.Hex())
 		return nil, err
 	}
 
+	if resBlock.Block == nil {
+		e.logger.Debugf("BlockByHash block not found for %s", hash.Hex())
```
Comment (fedekunze) [inline ethereum/rpc/backend/backend.go:142]:
```suggestion
		e.logger.Debugf("BlockByHash block not found", "hash", hash.Hex())
```

**Review by fedekunze — APPROVED** [review]

File: ethereum/rpc/backend/backend.go [hunk 6]
```diff
@@ -129,10 +134,15 @@ func (e *EVMBackend) GetBlockByNumber(blockNum types.BlockNumber, fullTx bool) (
 func (e *EVMBackend) GetBlockByHash(hash common.Hash, fullTx bool) (map[string]interface{}, error) {
 	resBlock, err := e.clientCtx.Client.BlockByHash(e.ctx, hash.Bytes())
 	if err != nil {
-		e.logger.Warningf("BlockByHash failed for %s", hash.Hex())
+		e.logger.WithError(err).Warningf("BlockByHash block not found", "hash", hash.Hex())
```
Comment (fedekunze) [inline ethereum/rpc/backend/backend.go:137]:
```suggestion
		e.logger.WithError(err).Debugln("BlockByHash block not found", "hash", hash.Hex())
```

File: ethereum/rpc/backend/backend.go [hunk 7]
```diff
@@ -129,10 +134,15 @@ func (e *EVMBackend) GetBlockByNumber(blockNum types.BlockNumber, fullTx bool) (
...
 		return nil, err
 	}
 
+	if resBlock.Block == nil {
+		e.logger.Debugf("BlockByHash block not found", "hash", hash.Hex())
```
Comment (fedekunze) [inline ethereum/rpc/backend/backend.go:142]:
```suggestion
		e.logger.Debugln("BlockByHash block not found", "hash", hash.Hex())
```

**Pull request merged**